### ISSUE TITLE
Issues using the accumulate tag

### ISSUE THREAD

**Issue by rossengeorgiev** [wazuh/wazuh-ruleset#123]
Hi, could you give me an example of how to use `accumulate` to combine multiple log lines into a single event?

**Comment by frgv** [comment]
Hello @rossengeorgiev 

`accumulate` tag is used for "remembering" stored data from a previous event with the same id (i.e. same session).

As an example, we can use openldap decoders, which implements `accumulate`: https://github.com/wazuh/wazuh-ruleset/blob/724c12c954505651276ec24a6a6e0b889c06c163/decoders/0185-openldap_decoders.xml

Certain user establishes a connection with the database:

`Apr 24 18:00:28 hostname slapd[20872]: conn=123456 fd=64 ACCEPT from IP=10.1.2.37:33957 (IP=10.1.2.2:389)`

This is the alert generated (`/var/ossec/logs/alerts/alerts.log`):


```
** Alert 1524585629.1520002: - syslog,access_control,
2018 Apr 24 18:00:28 manager1->/var/log/messages
Rule: 2508 (level 3) -> 'OpenLDAP connection open.'
Src IP: 10.1.2.37
Dst IP: 10.1.2.2
Apr 24 18:00:28 manager1 slapd: conn=123456 fd=64 ACCEPT from IP=10.1.2.37:33957 (IP=10.1.2.2:389)
```

After that, the user logs in as `example` user:

`Apr 24 18:00:41 hostname slapd[20872]: conn=123456 op=0 BIND dn="uid=example,ou=People,dc=example,dc=com" method=128`

No alert is generated because the rule associated to this event is level 0, but it would match this decoder:

```
<decoder name="openldap-bind">
    <parent>openldap</parent>
    <prematch>BIND </prematch>
    <regex>^conn=(\d+) op=\d+ BIND dn="\w+=(\w+),</regex>
    <order>id, dstuser</order>
    <accumulate/>
</decoder>
```

And, after that, for some reason, we receive another connection from the same user and same id (same session):

`Apr 24 18:00:54 hostname slapd[20872]: conn=123456 fd=64 ACCEPT from IP=10.1.2.37:33957 (IP=10.1.2.2:389)`


This time, because of the `<accumulate/>` tag usage, the alert generated is the following:

```
** Alert 1524585654.1520292: - syslog,access_control,
2018 Apr 24 18:00:54 manager1->/var/log/messages
Rule: 2508 (level 3) -> 'OpenLDAP connection open.'
Src IP: 10.1.2.37
Dst IP: 10.1.2.2
User: example
Apr 24 18:00:54 manager1 slapd: conn=123456 fd=64 ACCEPT from IP=10.1.2.37:33957 (IP=10.1.2.2:389)
```

As you can see, the field `User` is remembered from the last event received with the same id (123456), although that field is not present in the event that triggered the second alarm.

To recap, the `accumulate` tag can be used to keep track of previously decoded fields in events with same id.

I hope it helps, and please ask us any further question you may have.

Best regards, 

Fran G.

**Comment by rossengeorgiev** [comment]
Hi @frgv, is it possible to accumulate more than 2 lines? I have 4 separate lines, and I want to accumulate fields from all 4 lines.

**Comment by rossengeorgiev** [comment]
I am encountering two situation that don't work as I expect. The second case is not that important, but I think it is a clue to the issue. Perhaps accumulation works only on a couple of fields (e.g. srcip, srcuser, etc) and doesn't overwrite existing values.

Decoder:
```xml
<!--
Test Case 2: Combining fields from more than 2 lines

acctest id=1 user=wazuh
acctest id=1 action=add
acctest id=1 status=success (Expect: alert with fields user,action,status)

Test Case 2: Fields dont get overwriten for the same id

acctest id=1 user=wazuh
acctest id=1 user=root
acctest id=1 action=add (Expect: alert with fields user:root and action:add)
-->

<decoder name="acctest">
  <prematch>^acctest</prematch>
  <accumulate/>
</decoder>

<decoder name="acctest-user">
  <parent>acctest</parent>
  <prematch> user=</prematch>
  <regex>id=(\d+) user=(\S+)</regex>
  <order>id, dstuser</order>
  <accumulate/>
</decoder>

<decoder name="acctest-action">
  <parent>acctest</parent>
  <prematch> action=</prematch>
  <regex>id=(\d+) action=(\S+)</regex>
  <order>id, my_action</order>
  <accumulate/>
</decoder>

<decoder name="acctest-status">
  <parent>acctest</parent>
  <prematch> status=</prematch>
  <regex>id=(\d+) status=(\S+)</regex>
  <order>id, my_status</order>
  <accumulate/>
</decoder>
```

Rules:
```xml
<group name="acctest,">

  <rule id="999991" level="0">
    <decoded_as>acctest</decoded_as>
    <description>Accmulate test group</description>
  </rule>

  <rule id="999992" level="0">
    <if_sid>999991</if_sid>
    <match> user=</match>
    <description>Acctest user</description>
  </rule>

  <rule id="999993" level="10">
    <if_sid>999991</if_sid>
    <description>Acctest action</description>
  </rule>

  <rule id="999994" level="10">
    <if_sid>999991</if_sid>
    <description>Acctest status</description>
  </rule>

</group>
```

**Comment by rossengeorgiev** [comment]
I have an application that logs over a couple lines. I would like to accumulate initial fields like user, and some customs ones for them to be available when actions are performed. Then an alert could contain the user, customs fields together with the action. Naively, I thought the accumulate tag would be able to achieve that, but it seems the implementation is very limited and not generalized for versatility that decoders and rules provide. It would be nice to have this information documented in the reference.

**Comment by frgv** [comment]
Hi @rossengeorgiev 

Thanks very much for your feedback. As you correctly guessed, `accumulate` is not as flexible as it should be at this time. I just opened an issue (https://github.com/wazuh/wazuh/issues/576) for working on it.

Best regards,

Fran G.

**Comment by jesuslinares** [comment]
Thanks @rossengeorgiev for the feedback. Since there is an issue on Wazuh, I will close this one.

Regards,
Jesus.
